js-packages/@prestojs/viewmodel/src/ViewModelFactory.ts

```diff
@@ -0,0 +1,591 @@
+        const pkFieldNames = this._model.pkFieldNames;
+        const missing = pkFieldNames.filter(name => !(name in data));
+        const empty = pkFieldNames.filter(name => name in data && data[name] == null);
+        const errors: string[] = [];
+        if (empty.length > 0) {
+            errors.push(
+                `Primary key(s) '${empty.join("', '")}' was provided but was null or undefined`
+            );
+        }
+        if (missing.length > 0) {
+            errors.push(`Missing value(s) for primary key(s) '${missing.join("', '")}'`);
+        }
+
+        if (errors.length) {
+            throw new Error(errors.join(', '));
+        }
+
+        const assignedData: Record<string, any> = {};
+        const assignedFields: string[] = [];
+        const fields = this._model.fields;
+        for (const [key, value] of Object.entries(data)) {
+            const field = fields[key];
+            if (field) {
+                assignedData[key] = field.normalize(value);
+                assignedFields.push(key);
+            } else {
+                // TODO: Should extra keys in data be a warning or ignored?
+                console.warn(
+                    `Received value for key ${key}. No such field exists on ${this._model.name}`
+                );
+            }
+        }
+
+        this._assignedFields = assignedFields;
+        this._assignedFields.sort();
+        this._data = freezeObject(assignedData);
+
+        // Add getters for fields along with some traps for bad usage
+        // - If attempts to set a field throw an error in dev or log warning otherwise
+        // - If attempts to get a field that does exist but wasn't passed in `data` then
+        //   throw an error in dev or log a warning otherwise.
+        for (const fieldName of Object.keys(fields)) {
+            const definition: { set(value: any): any; get: () => any } = {
+                set(): void {
+                    const msg = `${fieldName} is read only`;
+                    if (isDev()) {
+                        throw new Error(msg);
+                    } else {
+                        console.warn(msg);
+                    }
+                },
+                get(): any {
+                    return assignedData[fieldName];
+                },
+            };
+            if (!this._assignedFields.includes(fieldName)) {
+                definition.get = (): void => {
+                    const msg = `'${fieldName}' accessed on ${
+                        this._model.name
+                    } but was not instantiated with it. Available fields are: ${this._assignedFields.join(
+                        ', '
+                    )}`;
+                    if (isDev()) {
+                        throw new Error(msg);
+                    } else {
+                        console.warn(msg);
+                    }
+                };
+            }
+            Object.defineProperty(this, fieldName, definition);
```

Comment:
A thought bubble: is it feasible to define everything as JS objects and use typescript's `readonly` to try to enforce this?

- regular JS, esp 3rd party libs (eg lodash) would still see properties as writable
- is it even possible in typescript to inject read-only properties into your new class based on the incoming field names?

Reply:
> A thought bubble: is it feasible to define field properties as regular JS object properties and use typescript's `readonly` to enforce immutability?
> 
> * regular JS, esp 3rd party libs (eg lodash) would still see properties as writable
> * is it even possible in typescript to inject read-only properties into your new class based on the incoming field names?

`readonly` doesn't change anything at runtime (if that's what you meant?). any properties added with `defineProperty` are by default immutable.

Reply:
Yeah, I was thinking about just letting TS enforce the readonly rather than the JS runtime.

Have thought a bit more about it and is probably going to be too dangerous as the consuming applications may not be TS and someone will end up modifying data directly and then complain about bugs in presto